File: leetcode/1477-find-two-non-overlapping-sub-arrays-each-with-target-sum/1477-find-two-non-overlapping-sub-arrays-each-with-target-sum.ts
/**
 * Finds the minimum sum of lengths of two non-overlapping subarrays that each sum to target.
 * If no such subarrays exist, returns -1.
 *
 * @param arr - The input array of numbers
 * @param target - The target sum for each subarray
 * @returns The minimum sum of lengths of two valid subarrays, or -1 if not possible
 */
function minSumOfLengths(arr: number[], target: number): number {
    // Map to store prefix sum -> index mapping
    // Key: prefix sum value, Value: index position
    const prefixSumToIndex = new Map<number, number>();
    prefixSumToIndex.set(0, 0);

    let currentPrefixSum = 0;
    const arrayLength = arr.length;

    // Array to store minimum length of valid subarray ending at or before index i
    // minLengthUpTo[i] represents the minimum length of a valid subarray in arr[0...i-1]
    const minLengthUpTo: number[] = Array(arrayLength + 1);

    const INFINITY = 1 << 30; // Large value representing infinity
    minLengthUpTo[0] = INFINITY;

    let minSumOfTwoLengths = INFINITY;

    // Iterate through the array to find valid subarrays
    for (let i = 1; i <= arrayLength; ++i) {
        const currentValue = arr[i - 1];
        currentPrefixSum += currentValue;

        // Inherit the minimum length from previous position
        minLengthUpTo[i] = minLengthUpTo[i - 1];

        // Check if there exists a subarray ending at current position with sum = target
        // We need prefixSum[j] such that prefixSum[i] - prefixSum[j] = target
        // Therefore, prefixSum[j] = prefixSum[i] - target
        if (prefixSumToIndex.has(currentPrefixSum - target)) {
            const startIndex = prefixSumToIndex.get(currentPrefixSum - target)!;
            const currentSubarrayLength = i - startIndex;

            // Update minimum length of valid subarray up to current position
            minLengthUpTo[i] = Math.min(minLengthUpTo[i], currentSubarrayLength);

            // Try to combine with the best subarray before startIndex
            // to get minimum sum of two non-overlapping subarrays
            minSumOfTwoLengths = Math.min(
                minSumOfTwoLengths,
                minLengthUpTo[startIndex] + currentSubarrayLength
            );
        }

        // Store current prefix sum and its index
        prefixSumToIndex.set(currentPrefixSum, i);
    }

    // If no valid pair of subarrays found, return -1
    return minSumOfTwoLengths > arrayLength ? -1 : minSumOfTwoLengths;
}

File: 1477-find-two-non-overlapping-sub-arrays-each-with-target-sum/1477-find-two-non-overlapping-sub-arrays-each-with-target-sum.java
class Solution {
    public int minSumOfLengths(int[] arr, int target) {
        // Map to store prefix sum -> index mapping
        // Key: prefix sum, Value: index (1-based)
        Map<Integer, Integer> prefixSumToIndex = new HashMap<>();
        prefixSumToIndex.put(0, 0); // Initialize with sum 0 at index 0

        int n = arr.length;

        // dp[i] stores the minimum length of a subarray with sum = target
        // ending at or before index i (1-based indexing)
        int[] minLengthUpToIndex = new int[n + 1];

        // Large value representing infinity for comparison
        final int INFINITY = 1 << 30;
        minLengthUpToIndex[0] = INFINITY; // No valid subarray before index 0

        int prefixSum = 0;
        int minTotalLength = INFINITY;

        // Iterate through the array (using 1-based indexing)
        for (int i = 1; i <= n; ++i) {
            int currentValue = arr[i - 1]; // Convert to 0-based array index
            prefixSum += currentValue;

            // Initially, inherit the minimum length from previous position
            minLengthUpToIndex[i] = minLengthUpToIndex[i - 1];

            // Check if we can form a subarray ending at current position with sum = target
            // We need prefixSum[j] = prefixSum[i] - target, where j < i
            if (prefixSumToIndex.containsKey(prefixSum - target)) {
                int startIndex = prefixSumToIndex.get(prefixSum - target);
                int currentSubarrayLength = i - startIndex;

                // Update the minimum length of subarray ending at or before index i
                minLengthUpToIndex[i] = Math.min(minLengthUpToIndex[i], currentSubarrayLength);

                // Try to combine with the best subarray before startIndex
                // to form two non-overlapping subarrays
                minTotalLength = Math.min(minTotalLength, minLengthUpToIndex[startIndex] + currentSubarrayLength);
            }

            // Store current prefix sum and its index
            prefixSumToIndex.put(prefixSum, i);
        }

        // If no valid pair of subarrays found, return -1
        return minTotalLength > n ? -1 : minTotalLength;
    }
}
